packages/core/src/storages/request_provider.ts

```diff
@@ -37,7 +37,68 @@ import { getRequestId, purgeDefaultStorages, QUERY_HEAD_MIN_LENGTH } from './uti
 
 export type RequestsLike = AsyncIterable<Source | string> | Iterable<Source | string> | (Source | string)[];
 
-export abstract class RequestProvider implements IStorage {
+/**
+ * Represents a provider of requests/URLs to crawl.
+ */
+export interface IRequestManager {
+    /**
+     * Returns `true` if all requests were already handled and there are no more left.
+     */
+    isFinished(): Promise<boolean>;
+
+    /**
+     * Resolves to `true` if the next call to {@apilink IRequestManager.fetchNextRequest} function
+     * would return `null`, otherwise it resolves to `false`.
+     * Note that even if the provider is empty, there might be some pending requests currently being processed.
+     */
+    isEmpty(): Promise<boolean>;
+
+    /**
+     * Returns number of handled requests.
+     */
+    handledCount(): Promise<number>;
+
+    /**
+     * Get the total number of requests known to the request manager.
+     */
+    getTotalCount(): number;
+
+    /**
+     * Get an offline approximation of the number of pending requests.
+     */
+    getPendingCount(): number;
```

Comment:
nits:
- Not sure if making all those async wouldn't make this more future-proof. 
- Also, for the next breaking release, we might want to rename those to follow the same naming convention? (all `get` prefix - or all no prefix)?

Reply:
Yeah, in v4, I would like to do `IRequestList -> IRequestLoader`, then `IRequestManager extends IRequestLoader` and get rid of `RequestProvider`. Making mostly everything `async` is probably a good idea.